### PR TITLE
Add docs covering compile time parameter usage

### PR DESCRIPTION
**Pull request by chrisroberts**

### CODE REVIEW

File: docs/compile-time-parameters.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+When compile time parameters are present SparkleFormation will automatically adjust the outputs
+of the template to include a `CompileState` output. The output will contain a JSON dump of the
+compile time parameter values used to generate the template. This allows update requests to
+fetch the previous state and seed the compilation of the updated template. Doing this provides consistency
+and removes any requirement of prior knowledge about how a template was used to build a stack.
+
+### Template Nesting
+
+Compile time parameters are local to a given template instance. A parent template can provide
+compile time parameters when nesting a template. This is done using the `:parameters` option
```
Comment (cwjohnston) [inline docs/compile-time-parameters.md]: Is the `compile_time_parameters` alias supported for nesting?

File: docs/compile-time-parameters.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+end
+~~~
+
+#### Template Modification
+
+When compile time parameters are present SparkleFormation will automatically adjust the outputs
+of the template to include a `CompileState` output. The output will contain a JSON dump of the
+compile time parameter values used to generate the template. This allows update requests to
+fetch the previous state and seed the compilation of the updated template. Doing this provides consistency
+and removes any requirement of prior knowledge about how a template was used to build a stack.
```
Comment (cwjohnston) [inline docs/compile-time-parameters.md]: maybe s/Doing this/This approach/ ?

File: docs/compile-time-parameters.md (hunk 1 above)
Comment (chrisroberts) [inline docs/compile-time-parameters.md]: No, it's just an alias for defining during initialization

File: docs/compile-time-parameters.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  end
+
+end
+~~~
+
+#### Parameter Declaration
+
+Declaring compile time parameters is done via the `:compile_time_parameters`
+option when instantiating a new template. This option expects a `Hash` value
+with the compile time parameter names as the key, and the parameter options
```
Comment (cwjohnston) [inline docs/compile-time-parameters.md]: I find this hard to read. Maybe "This option expects a `Hash` value in which the keys are parameter names, and their respective values are a `Hash` defining options for the parameter." ?

**Pull request merged**